File: dot_config/nvim/lua/plugins/ddc.lua
--[[
NOTE:
For some reason, ddc-ui-pum does not trigger User events...
So I prefer ddc-ui-native and use ddc-ui-pum only on cmdline.
--]]
local fn = vim.fn
local utils = require("atusy.utils")
local set_keymap = utils.set_keymap

local DEBUG = false

local function commandline_post_buf(buf, opts)
  if vim.api.nvim_buf_is_valid(buf) then
    vim.api.nvim_buf_call(buf, function()
      vim.fn["pum#set_option"]({ reversed = false })
      fn["ddc#custom#set_buffer"](opts or vim.empty_dict())
    end)
  end
end

local function commandline_pre(bufnr, mode)
  local opts = vim.fn["ddc#custom#get_buffer"]()
  vim.api.nvim_create_autocmd("User", {
    group = utils.augroup,
    pattern = "DDCCmdlineLeave",
    once = true,
    callback = function()
      commandline_post_buf(bufnr, opts)
    end,
  })
  vim.fn["pum#set_option"]({ reversed = true })
  fn["ddc#custom#patch_buffer"]("sourceOptions", {
    file = { forceCompletionPattern = [[(^e\s+|\S/\S*)]] },
    zsh = { enabledIf = [[getcmdline() =~# "^\\(!\\|Gin\\(Buffer\\)\\? \\)" ? v:true : v:false]] },
    -- ["_"] = mode == ":" and { keywordPattern = "[0-9a-zA-Z_:#-]*" },
  })

  -- Enable command line completion
  fn["ddc#enable_cmdline_completion"]()
end

local function setup()
  -- insert
  local patch_global = fn["ddc#custom#patch_global"]
  patch_global({
    ui = "pum",
    autoCompleteEvents = { "InsertEnter", "TextChangedI", "TextChangedP", "CmdlineEnter", "CmdlineChanged" },
    sources = { "nvim-lsp", "around", "file", "buffer", "skkeleton" },
    cmdlineSources = {
      [":"] = { "cmdline-history", "cmdline", "zsh", "file", "around" },
      ["@"] = { "cmdline-history", "input", "file", "around" },
      [">"] = { "cmdline-history", "input", "file", "around" },
      ["/"] = { "around", "line" },
      ["?"] = { "around", "line" },
      ["-"] = { "around", "line" },
      ["="] = { "input" },
    },
  })
  patch_global("sourceOptions", {
    around = { mark = "A" },
    buffer = { mark = "B" },
    ["nvim-lsp"] = { mark = "L", forceCompletionPattern = [[\S]] },
    file = {
      mark = "F",
      isVolatile = true,
      forceCompletionPattern = [[\S/\S*]],
    },
    input = { mark = "I", isVolatile = true, minAutoCompleteLength = 1 },
    ["cmdline"] = { mark = "C", minAutoCompleteLength = 1 },
    ["cmdline-history"] = {
      mark = "H",
      maxItems = 1,
      minAutoCompleteLength = 0,
      minKeywordLength = 2,
      sorters = {}, -- no sorters to prioritize latest history
    },
    zsh = {
      mark = "Z",
      minAutoCompleteLength = 0,
      minKeywordLength = 2,
    },
    ["_"] = {
      ignoreCase = true,
      matchers = { "matcher_fuzzy" },
      sorters = { "sorter_fuzzy" },
      converters = { "converter_fuzzy" },
    },
  })
  patch_global("sourceParams", {
    around = { maxSize = 500 },
    ["nvim-lsp"] = {
      enableResolveItem = true,
      enableAdditionalTextEdit = true,
    },
    zsh = { envs = { FPATH = vim.fn.system([[zsh -i -c 'echo -n "$FPATH"']]) } },
  })
  fn["ddc#custom#patch_filetype"]({ "bash", "zsh", "sh" }, {
    sources = { "nvim-lsp", "zsh", "around", "file", "buffer", "skkeleton" },
  })

  vim.keymap.set({ "i", "c" }, "<Tab>", function()
    if vim.fn["pum#visible"]() then
      return "<Cmd>call pum#map#insert_relative(+1)<CR>"
    end
    if vim.api.nvim_get_mode().mode == "c" then
      return vim.fn["ddc#map#manual_complete"]()
    end
    local col = fn.col(".")
    if col > 1 and string.match(fn.strpart(fn.getline("."), col - 2), "%s") == nil then
      return fn["ddc#map#manual_complete"]()
    end
    return "<Tab>"
  end, { expr = true })
  vim.keymap.set({ "i", "c" }, "<s-tab>", function()
    if vim.fn["pum#visible"]() then
      return "<Cmd>call pum#map#insert_relative(-1)<CR>"
    end
    return "<S-Tab>"
  end, { expr = true })
  vim.keymap.set({ "i", "c" }, "<c-y>", function()
    if vim.fn["pum#visible"]() then
      return "<Cmd>call pum#map#confirm()<CR>"
    end
    return "<C-Y>"
  end, { expr = true })
  vim.keymap.set("i", "<c-c>", function()
    if vim.fn["pum#visible"]() then
      return "<Cmd>call pum#map#cancel()<CR>"
    end
    return "<c-c>"
  end, { expr = true })

  -- cmdline
  set_keymap("n", "/", function()
    local ok, mes = pcall(commandline_pre, vim.api.nvim_get_current_buf())
    if DEBUG and not ok and mes then
      vim.notify(mes)
    end
    return "/"
  end, { expr = true })
  set_keymap({ "n", "x" }, ":", function()
    local ok, mes = pcall(commandline_pre, vim.api.nvim_get_current_buf(), ":")
    if DEBUG and not ok and mes then
      vim.notify(mes)
    end
    return ":"
  end, { expr = true })
  vim.keymap.set("c", "<c-c>", function()
    if vim.fn["pum#visible"]() then
      return "<Cmd>call pum#map#cancel()<CR>"
    end
    return "<c-u><bs>" -- Leave without histadd
  end, { expr = true })

  -- enable
  fn["popup_preview#enable"]()
  fn["ddc#enable"]()
end

return {
  {
    "Shougo/ddc.vim",
    -- dir = "/home/atusy/ghq/github.com/Shougo/ddc.vim",
    dependencies = {
      { "vim-denops/denops.vim" },
      { "Shougo/pum.vim" },
      { "Shougo/ddc-matcher_head" },
      { "Shougo/ddc-source-around" },
      { "Shougo/ddc-source-cmdline" },
      { "Shougo/ddc-source-cmdline-history" },
      { "Shougo/ddc-source-input" },
      { "Shougo/ddc-source-line" },
      { "Shougo/ddc-source-nvim-lsp" },
      { "Shougo/ddc-ui-pum" },
      { "matsui54/ddc-buffer" },
      { "LumaKernel/ddc-source-file" },
      { "Shougo/ddc-sorter_rank" },
      { "tani/ddc-fuzzy" },
      { "matsui54/denops-popup-preview.vim" },
      -- { "Shougo/ddc-converter_remove_overlap" },
    },
    config = setup,
  },
}

File: dot_config/nvim/lua/plugins/ddc/ddc.ts
import { BaseConfig } from "https://deno.land/x/ddc_vim@v3.9.0/types.ts";
import { ConfigArguments } from "https://deno.land/x/ddc_vim@v3.9.0/base/config.ts";
import { join } from "https://deno.land/std@0.196.0/path/mod.ts";

async function get_fpath() {
  const cmd = new Deno.Command("zsh", {
    args: ["-c", 'echo -n "$FPATH"'],
  });
  const output = await cmd.output();
  return (new TextDecoder()).decode(output.stdout);
}

export class Config extends BaseConfig {
  override async config(args: ConfigArguments): Promise<void> {
    const lazyroot = await args.denops.call(
      "luaeval",
      `require("lazy.core.config").options.root`,
    ) as string;

    const sources = [
      "lsp",
      "file",
      "around",
      "buffer",
      "skkeleton",
      "dictionary",
    ];

    ["zsh", "fish", "xonsh"].map((x) =>
      args.setAlias("source", x, "shell-native")
    );
    args.setAlias("source", "shell_history", "dictionary");
    args.setAlias("source", "ex_command_history", "cmdline-history");
    args.setAlias("source", "ex_command_history_cmd", "cmdline-history");
    args.setAlias("filter", "matcher_head_dictionary", "matcher_head");
    args.setAlias("filter", "matcher_head_shell_history", "matcher_head");
    args.setAlias("filter", "converter_ex_command", "converter_string_match");

    args.contextBuilder.patchGlobal({
      ui: "pum",
      backspaceCompletion: true, // NOTE: manual mentions occasional flickers
      autoCompleteEvents: [
        "InsertEnter",
        "TextChangedI",
        "TextChangedP",
        "CmdlineEnter",
        "CmdlineChanged",
        // "TextChangedT",
      ],
      sources: sources,
      cmdlineSources: {
        ":": [
          "ex_command_history_cmd",
          "fish",
          "zsh",
          "xonsh",
          "cmdline",
          "ex_command_history",
          "shell_history",
          "around",
        ],
        "@": ["input", "cmdline-history", "file", "around"],
        ">": ["input", "cmdline-history", "file", "around"],
        "/": ["around", "line"],
        "?": ["around", "line"],
        "-": ["around", "line"],
        "=": ["input"],
      },
      sourceOptions: {
        _: {
          ignoreCase: true,
          matchers: ["matcher_fuzzy"],
          sorters: ["sorter_fuzzy"],
          converters: ["converter_fuzzy"],
          timeout: 1000,
        },
        around: {
          mark: "A",
          converters: ["converter_fuzzy", "converter_dictionary"],
        },
        buffer: {
          mark: "B",
          converters: ["converter_fuzzy", "converter_dictionary"],
        },
        cmdline: {
          mark: "CMD",
          forceCompletionPattern: "\\S/\\S*|\\.\\w*",
        },
        dictionary: {
          mark: "Dict",
          matchers: ["matcher_head_dictionary", "matcher_fuzzy"],
          converters: ["converter_fuzzy", "converter_dictionary"],
          isVolatile: true,
          keywordPattern: "[a-zA-Z]+",
          maxItems: 30,
        },
        file: {
          mark: "F",
          isVolatile: true,
          forceCompletionPattern: "\\S/\\S*",
        },
        "cmdline-history": {
          mark: "HIST",
          minAutoCompleteLength: 0,
          minKeywordLength: 2,
          matchers: ["matcher_head"],
          sorters: [],
          converters: [],
        },
        input: {
          mark: "INPUT",
          forceCompletionPattern: "\\S/\\S*",
          isVolatile: true,
        },
        line: {
          mark: "LINE",
          matchers: ["matcher_vimregexp"],
          sorters: [],
          converters: ["converter_remove_overlap"],
        },
        "lsp": {
          mark: "L",
          forceCompletionPattern: "\\.\\w*|::\\w*|->\\w*",
          dup: "force",
        },
        "shell_history": {
          mark: "HIST_SH",
          matchers: ["matcher_head_dictionary", "matcher_fuzzy"],
          keywordPattern: "[^! ].*",
        },
        skkeleton: {
          mark: "SKK",
          matchers: ["skkeleton"],
          sorters: [],
          converters: [],
          minAutoCompleteLength: 2,
          isVolatile: true,
        },

        // aliases
        "ex_command_history_cmd": {
          mark: "RECENT",
          minAutoCompleteLength: 0,
          matchers: ["matcher_head"],
          sorters: [],
          converters: ["converter_ex_command"],
          enabledIf: "getcmdline() =~# ' ' ? v:false : v:true",
        },
        "ex_command_history": {
          mark: "HIST",
          minAutoCompleteLength: 0,
          matchers: ["matcher_head"],
          sorters: [],
          converters: [],
        },
        xonsh: {
          mark: "XONSH",
          isVolatile: true,
          minAutoCompleteLength: 0,
          minKeywordLength: 0,
        },
        fish: {
          mark: "FISH",
          isVolatile: true,
          minAutoCompleteLength: 0,
          minKeywordLength: 0,
        },
        zsh: {
          mark: "ZSH",
          isVolatile: true,
          minAutoCompleteLength: 0,
          minKeywordLength: 0,
        },
      },
      sourceParams: {
        around: { maxSize: 500 },
        buffer: {
          requireSameFiletype: false,
          limitBytes: 50000,
          fromAltBuf: true,
          forceCollect: true,
        },
        dictionary: {
          showMenu: false,
          smartCase: false,
          dictPaths: [join(lazyroot, "english-words/words_alpha.txt")],
        },
        file: {
          filenameChars: "[:keyword:].",
        },
        "lsp": {
          enableResolveItem: true,
          enableAdditionalTextEdit: true,
          confirmBehavior: "replace",
        },
        "shell_history": {
          showMenu: false,
          smartCase: false,
          dictPaths: ["/home/atusy/.zsh_history"],
        },

        // aliases
        fish: {
          shell: "fish",
          envs: {
            DDCVIM: "1",
            COLUMNS: "200", // to get more preview info
          },
        },
        zsh: {
          shell: "zsh",
          envs: {
            FPATH: await get_fpath(),
            COLUMNS: "200", // to get more preview info
          },
        },
        xonsh: {
          shell: "xonsh",
          envs: {
            COLUMNS: "200", // to get more preview info
          },
        },
      },
      filterParams: {
        // matcher
        matcher_head_dictionary: {
          maxMatchLength: 1,
        },
        matcher_head_shell_history: {
          maxMatchLength: 2,
        },
        // converter
        converter_dictionary: {
          dicts: [
            "kantan-ej-dictionary/kantan-ej-dictionary.json",
            "WebstersEnglishDictionary/dictionary.json",
          ].map((x) => join(lazyroot, x)),
        },
        converter_ex_command: {
          regexp: "[^ ]*[^ !]",
          convertAbbr: true,
        },
      },
    });

    const shellSources = ["fish", "zsh", "xonsh"]; // avoid shell_history as it can be too noisy
    ["sh", "bash", "fish", "xonsh", "zsh"].map(
      (x) => args.contextBuilder.patchFiletype(x, { sources: shellSources }),
    );
  }
}
